Link error for MsQuic+OpenSSL static build

### Discussed in https://github.com/microsoft/msquic/discussions/2142

<div type='discussions-op-text'>

<sup>Originally posted by **Wizmann** November 12, 2021</sup>
We come across a link error when we want to integrate MsQuic static library with our code, on Win10 21H1 with VS2019

The `msquic.lib` was successfully built with command like this: `scripts/build.ps1 -Config Release -Arch x64 -Static -Tls openssl -Clean`

But when we're trying to it with our code, it says: `warning LNK4098: defaultlib 'LIBCMT' conflicts with use of other libs;`

I thought it should not be a problem because we make that lib as `/NODEFAULTLIB` [here][1]. But it doesn't work as expected.

Could you please help to figure out the problem? Thanks!

[1]: https://github.com/microsoft/msquic/blob/cd8e64d1bacfa4925427a6599369136dd9ac7f1b/src/bin/CMakeLists.txt#L153</div>

--- a/src/bin/CMakeLists.txt
+++ b/src/bin/CMakeLists.txt
@@ -11,11 +11,11 @@ endif()
 
 if(BUILD_SHARED_LIBS)
     add_library(msquic SHARED ${SOURCES})
-    target_link_libraries(msquic PRIVATE core platform inc warnings logging base_link)
+    target_link_libraries(msquic PRIVATE core platform inc warnings logging base_link main_binary_link_args)
     set_target_properties(msquic PROPERTIES OUTPUT_NAME ${QUIC_LIBRARY_NAME})
 else()
     add_library(msquic_static STATIC static/empty.c)
-    target_link_libraries(msquic_static PRIVATE core platform inc logging)
+    target_link_libraries(msquic_static PRIVATE core platform inc logging main_binary_link_args)
     target_compile_definitions(msquic_static PUBLIC QUIC_BUILD_STATIC)
     set_property(TARGET msquic_static PROPERTY FOLDER "${QUIC_FOLDER_PREFIX}libraries")
 
@@ -150,7 +150,7 @@ else()
 
         # Exclude libraries for the corresponding msvcrts
         # https://docs.microsoft.com/en-us/cpp/error-messages/tool-errors/linker-tools-warning-lnk4098?view=msvc-160
-        string(APPEND NODEFAULTS "/NODEFAULTLIB:msvcrt.lib /NODEFAULTLIB:msvcrtd.lib $<IF:$<CONFIG:Debug>,/NODEFAULTLIB:libcmt.lib /NODEFAULTLIB:msvcrtd.lib,/NODEFAULTLIB:libcmtd.lib>")
+        string(APPEND NODEFAULTS "/NODEFAULTLIB:msvcrt.lib /NODEFAULTLIB:msvcrtd.lib /NODEFAULTLIB:libcmt.lib /NODEFAULTLIB:libcmtd.lib")
         # Uncomment to inspect which libraries will have referenced symbols ignored
         # message(STATUS "NODEFAULTS: ${NODEFAULTS}")
 

--- a/src/core/CMakeLists.txt
+++ b/src/core/CMakeLists.txt
@@ -62,7 +62,7 @@ endif()
 
 target_link_libraries(core PUBLIC inc)
 
-target_link_libraries(core PRIVATE warnings)
+target_link_libraries(core PRIVATE warnings main_binary_link_args)
 
 set_property(TARGET core PROPERTY FOLDER "${QUIC_FOLDER_PREFIX}libraries")
 

--- a/src/inc/CMakeLists.txt
+++ b/src/inc/CMakeLists.txt
@@ -47,3 +47,8 @@ endif()
 
 add_library(warnings INTERFACE)
 target_compile_options(warnings INTERFACE ${QUIC_WARNING_FLAGS})
+
+add_library(main_binary_link_args INTERFACE)
+if (WIN32 AND NOT BUILD_SHARED_LIBS)
+    target_compile_options(main_binary_link_args INTERFACE /Zl)
+endif()

--- a/src/platform/CMakeLists.txt
+++ b/src/platform/CMakeLists.txt
@@ -43,7 +43,7 @@ add_library(platform STATIC ${SOURCES})
 
 target_link_libraries(platform PUBLIC inc)
 
-target_link_libraries(platform PRIVATE warnings)
+target_link_libraries(platform PRIVATE warnings main_binary_link_args)
 
 set_property(TARGET platform PROPERTY FOLDER "${QUIC_FOLDER_PREFIX}libraries")
 

